### PR TITLE
add `--hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check` argument for holding out part of train split as additional generalization check after choosing early stopping style model based on validation performance (does not change train/valid/test split, just reallocates part of train split for additional check)

### PR DESCRIPTION
**Pull request by willy-b**

### CODE REVIEW

**Comment by willy-b** [comment]
Note, in most cases this option is not useful without the changes coming in https://github.com/willy-b/tiny-GIN-for-PCBA-577-WNV/pull/10 .

File: main_gin.py [hunk 1]
```diff
@@ -99,8 +100,21 @@ def set_seeds(seed):
...
-valid_loader = DataLoader(dataset[split_idx["valid"]], batch_size=config["batch_size"], shuffle=False)
+
+# note they are preshuffled and presplit into training/validation/test splits
+if args.hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check and (len(split_idx["train"]) < 2 * len(split_idx["valid"])):
+    raise Exception("Cannot use --hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check when validation set is greater than half the train set size due to insufficient data remaining.")
+
+train_split = split_idx["train"] if not args.hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check else split_idx["train"][len(split_idx["valid"]):]
+valid_split = split_idx["valid"]
+
+dev_split = None if not args.hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check else split_idx["train"][:len(split_idx["valid"])]# NOT TEST, this is taking some of the existing training split
```
Comment (willy-b) [inline main_gin.py:111]: Note, the `--hold_out_addl_data_from_train_set_as_dev_for_addl_generalization_check` argument as added here would NOT ensure that the new split out of training set would contain any minimum number of the active molecules (they make up less than 0.2% of the examples in this data). For WNV PCBA 577 there are VERY FEW active molecules to split between the different sets and if this is not checked I was observing e.g. 7 molecules only in this hold out leading to really unreliable (high variance) assessments of generalization in this training data subset hold out. So making a change to ensure that the holdout from train of the same size as the validation set has the same number of active molecules as the validation set for fair comparison.

Comment (willy-b) [reply main_gin.py:111]: See https://github.com/willy-b/tiny-GIN-for-PCBA-577-WNV/pull/10/files#r1797735363 for upcoming change.
